test-data/unit/check-newsemanal.test

```diff
@@ -3241,3 +3241,52 @@ class b:
 x = x[1]  # E: Cannot resolve name "x" (possible cyclic definition)
 y = 1[y]  # E: Value of type "int" is not indexable \
           # E: Cannot determine type of "y"
+
+[case testBuiltinTypeType]
+# flags: --python-version 3.10 --disallow-any-generics
+
+from typing import Any, cast, Union, TypeVar
+from typing_extensions import TypeAlias
+
+a: type  # E: Missing type parameters for generic type "type"
+reveal_type(a)  # N: Revealed type is "Type[Any]"
+
+b: Any
+c = cast(type, b)  # E: Missing type parameters for generic type "type"
+reveal_type(c)  # N: Revealed type is "Type[Any]"
+
+d: list[type]  # E: Missing type parameters for generic type "type"
+reveal_type(d)  # N: Revealed type is "builtins.list[Type[Any]]"
+
+class E(list[type]): ...  # E: Missing type parameters for generic type "type"
+reveal_type(E()[0])  # N: Revealed type is "Type[Any]"
+
+class F(tuple[type]): ...  # E: Missing type parameters for generic type "type"
+reveal_type(F()[0])  # N: Revealed type is "Type[Any]"
+
+class G(tuple[list[Union[type, int]], ...]): ...  # E: Missing type parameters for generic type "type"
+reveal_type(G()[0])  # N: Revealed type is "builtins.list[Union[Type[Any], builtins.int]]"
+
+h: Union[type, int]  # E: Missing type parameters for generic type "type"
+reveal_type(h)  # N: Revealed type is "Union[Type[Any], builtins.int]"
+
+i: type | int  # E: Missing type parameters for generic type "type"
+reveal_type(i)  # N: Revealed type is "Union[Type[Any], builtins.int]"
+
+j: TypeAlias = type  # E: Missing type parameters for generic type "type"
+class K(metaclass=j): ...  # E: Invalid metaclass "j"
+l: j
+reveal_type(l)  # N: Revealed type is "Type[Any]"
+
+m = type
+class N(metaclass=m): ...
+
+O = TypeVar("O", bound=type)  # E: Missing type parameters for generic type "type"
+def p(q: O) -> O: ...
+reveal_type(p(str))  # N: Revealed type is "def () -> builtins.str"
```

Comment:
Add a test that a non-type is rejected, e.g. `p(1)  # E: <something about incompatible types>`

Reply:
Okay, I add `p(1)  # E: Value of type variable "O" of "p" cannot be "int"`.